Update Nest configuration flow to handle upcoming changes to Pub/Sub provisioning

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update Nest configuration flow to handle an upcoming change by Nest to move to user provisioned Pub/Sub topics. Existing Nest Device Access projects are unaffected.

Background: Currently when a user sets up the device access console, Nest provisions a Pub/Sub topic automatically. The user must must create a Subscription, which is currently handled by nest. I am working closely with the Nest engineering team while they are changing the process for how this works.

Updated flow: 
1. When setting up the Device Access console, users need to manually create a Pub/Sub topic, configure the permissions, and enter it in the Device Access console.
2. The Device Access console will verify it is configured correctly.
3.  The new configuration flow in this PR will allow for these various user states:
    - Topic: The old way, where Nest has configured the Pub/Sub topic
    - Topic: The new way, where the user has already created a Pub/Sub topic
    - Subscription: No existing subscription exists for the topic, and it can be created
    - Subscription: The user or home-assistant has already created a subscription for the Topic.

This change does the minimal set up to support these old and new new flow combinations. We may allow further changes or clarifications (e.g. Pub/Sub topic provisioning as well) as the change is rolled out. 

Rollout:
- We will first launch this for SDM managed topics, then update the documentation for user provisioned topics once that is launched in production.
- I've tested both scenarios manually using a pre-prod environment.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

Topic selection for user created topics:
<img width="588" alt="Subscriber Setup" src="https://github.com/user-attachments/assets/dbd46ecf-4e54-4693-bd73-a5f2876ce137">

Subscriber selection for user created subscriptions:
<img width="591" alt="Topic setup" src="https://github.com/user-attachments/assets/197b6afa-dd57-41c5-a950-a80cf28d5d49">




## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

allenporter committed:
Update Nest configuration flow to allow for user managed Pub/Sub topics

allenporter committed:
Update translations, apis, and error handling

allenporter committed:
Update translations

allenporter committed:
Mark existing subscriptions as imported so they are not removed on integration removal

joostlek commented on homeassistant/components/nest/config_flow.py:
```diff
@@ -281,23 +366,39 @@ async def async_step_pubsub(
+            )
+        subscriptions[CREATE_NEW_SUBSCRIPTION_KEY] = "Create New"
         return self.async_show_form(
-            step_id="pubsub",
+            step_id="pubsub_subscription",
             data_schema=vol.Schema(
                 {
-                    vol.Required(CONF_CLOUD_PROJECT_ID, default=cloud_project_id): str,
+                    vol.Optional(
+                        CONF_SUBSCRIPTION_NAME,
+                        default=next(iter(subscriptions)),
+                    ): vol.In(subscriptions),
                 }
             ),
```
I am not 100% sure on this, but you can pass a translation key to a selectselector. maybe we can use that to translate the create_new. But I am not sure if all keys that are in the options need to be in string.json, because ideally it shows the raw strings if its not found.

Not a blocker, just curious

allenporter replied:
Yeah, I was thinking the same thing that it would be ideal to translate create new, but then still have it work for th other options.

```
Please select a topic:
- ( ) Use A
- ( ) Use B
- ( ) Use C
- ( ) Create a new topic
```

My assumption was that given some of the work to enforce translation keys existing, that we wouldn't want to do this. I will follow up and get a definitive answer though.

joostlek commented on homeassistant/components/nest/strings.json:
```diff
@@ -26,12 +26,26 @@
         }
       },
+      "pubsub_topic": {
+        "title": "Configure Cloud Pub/Sub Topic",
+        "description": "Nest devices publish updates on a Cloud Pub/Sub Topic. Select the Pub/Sub topic below that is the same as the [Device Access Console]({device_access_console_url}). See the integration documentation for [more info]({more_info_url}).",
+        "data": {
+          "topic_name": "Pub/Sub Topic Name"
```
is `Topic` a brand name or is it just a normal word, otherwise we should lowercase that word

allenporter replied:
Lowercasing.

joostlek commented on homeassistant/components/nest/strings.json:
```diff
@@ -26,12 +26,26 @@
+          "topic_name": "Pub/Sub Topic Name"
+        }
+      },
+      "pubsub_subscription": {
+        "title": "Configure Cloud Pub/Sub Subscription",
+        "description": "Home Assistant receives realtime Nest device updates with a Cloud Pub/Sub Subscription for topic `{topic}`.\n\nSelect an existing subscription below if one already exists, or the next step will create a new one for you. See the integration documentation for [more info]({more_info_url}).",
+        "data": {
+          "subscription_name": "Pub/Sub Subscription Name"
+        }
```
idem for subscription name

allenporter replied:
Lowercasing.

joostlek reviewed (changes requested):
Reading through this, you are much more in depth of this, but it looks quite logical. So awesome! I do have a few comments/remarks.

Also, I am now wondering, does `nest` support multiple config entries and could they collide when you for some reason select the same pubsub? (Just a random though)

allenporter committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

allenporter commented on homeassistant/components/nest/api.py:
```diff
@@ -111,29 +114,52 @@ async def new_subscriber(
     ):
         raise TypeError(f"Unexpected auth implementation {implementation}")
-    if not (subscriber_id := entry.data.get(CONF_SUBSCRIBER_ID)):
-        raise ValueError("Configuration option 'subscriber_id' missing")
+    if not (
+        subscription_name := entry.data.get(
+            CONF_SUBSCRIPTION_NAME, entry.data.get(CONF_SUBSCRIBER_ID)
+        )
+    ):
+        raise ValueError(
+            "Configuration option 'subscription_name' or 'subscriber_id' missing"
+        )
```
The first one is for typing:
```
homeassistant/components/nest/api.py:119: error: "AbstractOAuth2Implementation" has no attribute "client_id"  [attr-defined]
homeassistant/components/nest/api.py:120: error: "AbstractOAuth2Implementation" has no attribute "client_secret"  [attr-defined]
```
so we could cast instead, but that's why its here.

The second one also shouldn't happen. I think a couple years ago there were a large number of possible states for nest configurations to be in (yaml, multiple auth types, imported subscriptions, manually created subscriptions, etc) that I was probably being paranoid and wanting a clear error message. I've updated this to assume the key exists.

allenporter replied:
Sent https://github.com/home-assistant/core/pull/130057 and added a test case that was missing from this PR to exercise the bug.

allenporter commented on homeassistant/components/nest/config_flow.py:
```diff
@@ -238,40 +243,120 @@ async def async_step_device_project(
+                        topic_name,
+                        subscription_name,
+                    )
+                    _LOGGER.debug("Created?")
```
Removing this instead, didn't meant to include this in the PR.

allenporter committed:
Fix translations, typing, and remove debug statements

allenporter commented:
> Also, I am now wondering, does `nest` support multiple config entries and could they collide when you for some reason select the same pubsub? (Just a random though)

Yes, nest does support multiple config entries. The main thing that users may mess up here is picking a pubsub topic that is not configured with the nest device access project.  That is, the main steps need to be:
- Create a topic
- Create a subscription for that topic
- Ensure Nest Device Access console the name of the topic
- Ensure Home Assistant has the name of the subscription

(Aside: Home Assistant asks for a topic name and a subscription in the config flow steps to help create them and to help the user know they need to do (3) and manually verify the device access console is set up correctly.)

So the main thing that can get messed up is that the nest device access console is configured with the wrong topic, then no events will be received. It's not just that you configure two with the same subscription, its that it implies one is wrong.  So we could ensure that the user doesn't configure the same subscription twice since it indicates something else is wrong?

allenporter committed:
Remove unnecessary test for empty subscriber

allenporter commented:
I fixed the failing test that was covering the missing field case I removed. This is now ready for review, and i'd love to get into Nov release for Nest team's migration.
